File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
function solution(jobs) {
    var answer = 0;
    const PQ = [];
    let time=0, i=0;
    
    jobs.sort((a,b)=>a[0]-b[0]);
    
    while(PQ.length > 0 || i < jobs.length){
        if(jobs.length > i && jobs[i][0] <= time){
            PQ.push(jobs[i++]);
            PQ.sort((a,b)=>a[1]-b[1]);
            continue;
        }
        
        
        if(PQ.length > 0){
            const [start, work] = PQ.shift();
            answer += time - start + work;
            time += work;
        }
        
        else if(PQ.length == 0){
            time = jobs[i][0];
        }
    }
    
    
    
    return parseInt(answer/jobs.length);
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    answer = 0
    
    heap = []
    jobs.sort(key=lambda x: x[0])
    idx = 0
    time = 0
    
    while idx < len(jobs) or heap:
        while idx < len(jobs) and jobs[idx][0] <= time: 
            start, duration = jobs[idx]
            heapq.heappush(heap ,(duration, start))
            idx += 1
        
        if heap:
            duration, start = heapq.heappop(heap)
            time += duration
            answer += time - start
        
        else:
            time = jobs[idx][0]

        
    return answer // len(jobs)